src/Menu/useMenu.ts

```diff
@@ -0,0 +1,281 @@
+         close,
+      };
+   }, [
+      menuButtonId,
+      menuId,
+      menuButtonRef,
+      menuRef,
+      openAndFocusFirstItem,
+      openAndFocusLastItem,
+      isOpen,
+      focusContext,
+      close,
+   ]);
+
+   return context;
+}
+
+export type MenuListProps = React.HTMLAttributes<HTMLUListElement>;
+
+type MenuListInternalProps = ElementProps<HTMLUListElement> & {
+   'data-active': boolean;
+};
+
+export function useMenuList(props: MenuListProps): MenuListInternalProps {
+   const { menuRef, isOpen, menuId, close, focusContext } = useMenuContext();
+
+   const onKeyDown = React.useCallback(
+      (event: React.KeyboardEvent) => {
+         const eventKey = normalizeEventKey(event);
+
+         const keyMap: { [key: string]: (event: React.KeyboardEvent) => void } = {
+            Tab: (event) => {
+               event.preventDefault();
+            },
+            Escape: close,
+            ArrowDown: () => {
+               focusContext.focusNext();
+            },
+            ArrowUp: () => {
+               focusContext.focusPrevious();
+            },
+         };
+
+         const action = keyMap[eventKey];
+
+         if (action) {
+            event.preventDefault();
+            action(event);
+            return;
+         }
+      },
+      [focusContext, close]
+   );
+
+   const menuListProps = React.useMemo((): MenuListInternalProps => {
+      return {
+         ...props,
+         id: menuId,
+         role: 'menu',
+         'aria-orientation': 'vertical',
+         'data-active': isOpen,
+         tabIndex: -1,
+         ref: menuRef,
+         onKeyDown,
+      };
+   }, [props, menuRef, isOpen, onKeyDown]);
+   return menuListProps;
+}
+
+export type MenuItemProps = React.PropsWithChildren<{
+   className?: string;
+   onSelect?: (event: React.MouseEvent<HTMLLIElement, MouseEvent> | React.KeyboardEvent) => void;
+   url?: string;
+}>;
+
+export function useMenuItem({ onSelect, url, ...props }: MenuItemProps) {
+   const { isOpen, focusContext } = useMenuContext();
+   const itemRef = React.useRef<any>(null);
+
+   const { isFocused, index } = useFocusItem({
+      isMenuOpen: isOpen,
+      context: focusContext,
+      ref: itemRef,
+   });
+
+   const onKeyDown = React.useCallback(
+      (event: React.KeyboardEvent) => {
+         const eventKey = normalizeEventKey(event);
+         if (eventKey === 'Enter' && url == null) {
+            // event.preventDefault();
+            onSelect?.(event);
+         }
+      },
+      [onSelect, url]
+   );
+
+   const onMouseEnter = React.useCallback(() => {
+      focusContext.focus(index);
+   }, [focusContext.focus, itemRef, index]);
+
+   const onMouseMove = React.useCallback(() => {
+      if (document.activeElement !== itemRef.current) {
+         focusContext.focus(index);
+      }
+   }, [focusContext.focus, itemRef, index]);
+
+   const onMouseLeave = React.useCallback(() => {
+      focusContext.blur();
+   }, [focusContext.blur]);
```

Comment:
**TLDR**: I think this provides possibly the best usability and accessibility, addressing a many edge cases.

---
These are very subtle details, and what they do it's best seen by using the menu.
Basically to make a menu component accessible, you need to handle focus yourself, because the browser doesn't do it right for custom components (here is a reference to the WAI-ARIA [menu button](https://www.w3.org/TR/wai-aria-practices-1.1/#menu) design pattern). 

When you open the menu, you want the first item to be focused so that it can be selected simply by pressing enter on the keyboard. 
When you hover a menu item with the mouse, you want that item to be in  an hover state, to signal that it can be selected. Now, the reason why you want to track mouse events, is because you want the item not only to be hovered, but also to be focused, so that if the user presses enter on the keyboard, that item is selected. 

`onMouseMove` is used to address this edge case: 
- User hovers an item with the mouse -> the item is focused
- User moves to other items using arrow keys -> focus moves accordingly
- User interacts with the mouse again (over the same initially hovered item, this can be detected only with `onMouseMove`) -> focus is moved back to that item

I share your concern about the performance of this, but I think that unless we prove that this effectively is an issue there's no reason to compromise. Using it I don't notice any lag or performance impact. As an optimization we could throttle the handler, but I wouldn't do it unless proven necessary.

Reply:
I appreciate the explanation and your commitment to accessibility.